Add rich text authoring to several interactives

kswenson opened the pull request:
Rich text added to:
- Open Response prompt, hint and feedback
- Multiple Choice prompt, hint and feedback
- Scaffolded Question prompt and hint
- Fill in the Blank prompt and hint

To enable for `prompt`, simply add the following to the interactive schema:
```
uiSchema: {
  prompt: {
    "ui:widget": "richtext"
  }
}
```

Requires that [LARA #578](https://github.com/concord-consortium/lara/pull/578) be merged for rich text hints to work as expected.

Notes:
- Implementation requires the use of React's `dangerouslySetInnerHTML()` property to install the authored rich text.
- Enzyme doesn't work with content rendered with `dangerouslySetInnerHTML()`, so some unit tests now use the [@testing-library/react](https://testing-library.com/docs/react-testing-library/intro).
- Cypress' `type()` function doesn't work on the rich text editor, so I had to disable some of the cypress tests. I spent a little time trying to find an alternative that worked without success.
- `webpack.config.js` has been modified to support `.global.scss` files as CSS files that should not be processed as CSS modules.
- There are likely to be merge conflicts with @pjanik 's [logging PR](https://github.com/concord-consortium/question-interactives/pull/32), so that should probably be merged first.

kswenson commented on src/fill-in-the-blank/components/runtime.test.tsx:
```diff
@@ -1,110 +1,168 @@
     expect(newState).toEqual({
       answerType: "interactive_state",
-      answerText: "Test prompt with [ Test response ] and [ New response ].",
+      answerText: "<p>Test prompt with [ Test response ] and [ New response ].</p>",
```
>It'd be nice if Cypress tests don't stay like that forever. ;) But I guess they can be fixed in other PR, depends on timing and priorities. Probably some PT story would be nice so it doesn't get lost.

Agreed, but I don't know how much work is involved in getting cypress tests to work with the RTE so it seems reasonable to put it off to another PR given this one has already gotten quite large.

>The second thing would be HTML in `answerText` - I feel like we should avoid it at this point, but maybe @scytacki has some opinion on that too (Teacher Report could render HTML if we want, and it could have some benefits too).

Good point -- I suspect you're right. We'd have to strip the tags from the prompt in that case.

>I'm wondering about using `dangerouslySetInnerHTML()` everywhere. Aren't there available some libraries that parse HTML and generate React components? I could imagine that they could also deal with incorrect HTML and escape dangerous parts (since Runtime doesn't know where the HTML is coming from in fact). I googled quickly and saw: https://www.npmjs.com/package/html-react-parser . This could also fix problems with Enzyme.

I think that's worth an experiment. As you say it would be nice not to have to use `dangerouslySetInnerHTML()`. It looks like html-react-parser doesn't specifically deal with security issues. I wondered about using `DOMPurify` on the interactive side as well but the risks seemed much less in an interactive iframe than on the LARA side, for instance, so I wasn't sure if it was worth it. I think I prefer @testing-library/react at this point, however, so I'm not switching the tests back to Enzyme even if it works. 😉

kswenson replied:
I've added some additional commits which:
- convert the `answerText` to plain text
- replace use of `dangerouslySetInnerHTML()` with `html-react-parser` and `DOMPurify`

kswenson commented on src/shared/widgets/rich-text/rich-text-widget.tsx:
```diff
@@ -0,0 +1,70 @@
+
+  const handleFocus = useCallback(() => {
+    props.onFocus(props.id, slateToHtml(valueRef.current));
+  }, [valueRef.current]);
```
I'm not a big fan of `useCallback()` either and I'm certainly not using it for performance reasons. `handleFocus()` and `handleBlur()` are called asynchronously and this was the only way I could get them not to close on a stale state value. I may play around with it a bit more to see if there's another solution, but this was simplest solution I found that actually worked.

kswenson replied:
I just tried this again and it turns out the the `useCallback()` isn't required. I thought I had tried that combination. The `useRefState()` is required, however, to prevent the stale closure. I've added a commit with the change.

pjanik replied:
I don't think it's worth investigating, but in this case, it seems to me that this closure needs to happen somewhere in Slate Editor code. Maybe it doesn't update these event handlers once they're assigned or something like that? Now, we have regular functions in the top-level of the render function that should get recreated on every state update (including `value`).

pjanik commented:
Thanks for updates, it looks good to me.

kswenson commented:
@scytacki @pjanik I fixed the cypress tests. It turned out that the problem wasn't that typing wasn't working but that `blur()`ing wasn't working. I had to add a plugin to simulate a tab to trigger the blur. Two small mysteries remain:

1. I had to add a wait between the blur event and the testing for the received `setAuthoredState()` message contents.  Without it the call to `cy.window().should("have.property", "receivedMessage")` would consistently pass but then the received message would be `undefined`.
1. There's one final test that remains commented out -- the test that verifies that in report mode the `<input>`s are readonly/disabled. For some reason cypress is able to change the contents now when it couldn't before. I've verified that the `readonly` and `disabled` attributes are being applied correctly. I've stopped in the debugger and looked at the `<input>` before and after and they appear identical. I've added explicit tests for the `readonly` and `disabled` attributes (even though only the latter should be necessary) which I'm inclined to think should be good enough.

My inclination is not to hold up this PR any further for either of these mysteries.

scytacki commented:
Looks good to me. I'm going to merge this.
